Prevent automatic rejection of confirmations

Gudahtt opened the pull request:
Confirmations are now only automatically rejected if a user explicitly closes the notification window. If we close the window programmatically because there are no notifications left to show, nothing gets rejected.

This partially avoids a race condition where a confirmation gets rejected automatically without the user having seen the confirmation first. This could happen if the confirmation was processed just as the notification window was being closed.

It's still possible for a confirmation that the user has never seen to get rejected as a result of the user closing the window. But at least now it's no longer possible for a confirmation to get rejected in this manner after the user resolves the last confirmation in the queue.

If there is a queued confirmation when the popup is automatically closed, the popup is triggered again so that the user is presented with it.

Gudahtt committed:
Prevent automatic rejection of confirmations

Confirmations are now only automatically rejected if a user explicitly
closes the notification window. If we close the window programmatically
because there are no notifications left to show, nothing gets rejected.

This partially avoids a race condition where a confirmation gets
rejected automatically without the user having seen the confirmation
first. This could happen if the confirmation was processed just as the
notification window was being closed.

It's still possible for a confirmation that the user has never seen to
get rejected as a result of the user closing the window. But at least
now it's no longer possible for a confirmation to get rejected in this
manner after the user resolves the last confirmation in the queue.

Gudahtt commented:
This depends upon #13202

rekmarks commented on app/scripts/lib/notification-manager.js:
```diff
@@ -72,7 +82,10 @@ export default class NotificationManager extends EventEmitter {
     if (windowId === this._popupId) {
       this._popupId = undefined;
-      this.emit(NOTIFICATION_MANAGER_EVENTS.POPUP_CLOSED);
+      this._popupAutomaticallyClosed = undefined;
```
Why not `false`?

rekmarks replied:
Also, doesn't this make it so that `automaticallyClosed` is always `undefined`?

Gudahtt replied:
Oops! Yes, good catch.

`false` works as well I suppose. I left it unassigned by default because when this isn't set, we don't know whether the next "window close" event is automatic or not. The options are "unknown" or "closed automatically", so `undefined` felt more appropriate

Gudahtt replied:
Fixed!

rekmarks commented on ui/pages/home/home.component.js:
```diff
@@ -46,6 +46,17 @@ const LEGACY_WEB3_URL =
+    totalUnapprovedCount === 0 &&
+    !isSigningQRHardwareTransaction
+  );
+}
```
nit: breathing space
```suggestion
}

```

Gudahtt replied:
Done!

rekmarks reviewed (commented):
Looks good overall!

Gudahtt committed:
Fix bug that prevented automatic closure detection

All windows were being detected as explicit window closures,
essentially just as they were previously, because this variable was
cleared too soon.

Gudahtt committed:
Re-open popup when necessary

After the window is automatically closed, a confirmation may have been
queued up while the window was closing. If so, the popup is now re-
opened.

Gudahtt committed:
Add newline between function and class

Gudahtt commented:
When testing this with the `snaps` branch, it appeared that the problem was fixed but there was a new problem now: the popup never opened. I realized that it's because if the confirmation was queued while the window was closing, it would never be shown.

I've updated this branch to automatically re-trigger the UI in this circumstance as well. The PR description has been updated accordingly:
>If there is a queued confirmation when the popup is automatically closed, the popup is triggered again so that the user is presented with it.

rekmarks reviewed (approved):
LGTM!
